src/Raven.Server/ServerWide/ClusterStateMachine.License.cs

```diff
@@ -0,0 +1,206 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using Raven.Client.Exceptions.Commercial;
+using Raven.Client.ServerWide;
+using Raven.Server.Commercial;
+using Raven.Server.Json;
+using Raven.Server.ServerWide.Commands;
+using Raven.Server.ServerWide.Commands.ETL;
+using Raven.Server.ServerWide.Commands.Indexes;
+using Raven.Server.ServerWide.Commands.PeriodicBackup;
+using Raven.Server.ServerWide.Context;
+
+namespace Raven.Server.ServerWide;
+
+public sealed partial class ClusterStateMachine
+{
+    private const int MinBuildVersion54118 = 54_118;
+
+    private static readonly List<string> _licenseLimitsCommandsForCreateDatabase = new()
+    {
+        nameof(PutIndexCommand),
+        nameof(PutIndexesCommand),
+        nameof(UpdatePeriodicBackupCommand),
+        nameof(UpdatePullReplicationAsSinkCommand),
+        nameof(UpdatePullReplicationAsHubCommand),
+        nameof(UpdateExternalReplicationCommand),
+        nameof(AddRavenEtlCommand),
+        nameof(AddSqlEtlCommand),
+        nameof(EditTimeSeriesConfigurationCommand),
+        nameof(EditDocumentsCompressionCommand),
+        nameof(AddElasticSearchEtlCommand),
+        nameof(AddOlapEtlCommand),
+        nameof(AddQueueEtlCommand)
+    };
+
+    private void AssertLicenseLimits(string type, ServerStore serverStore, DatabaseRecord databaseRecord, ClusterOperationContext context)
+    {
+        switch (type)
+        {
+            case nameof(PutIndexCommand):
+            case nameof(PutIndexesCommand):
+                if (CanAssertLicenseLimits(context, minBuildVersion: MinBuildVersion54118) == false)
+                    return;
+
+                if (serverStore.LicenseManager.LicenseStatus.HasAdditionalAssembliesFromNuGet)
+                    return;
+
+                if (LicenseManager.HasAdditionalAssembliesFromNuGet(databaseRecord.Indexes) == false)
+                    return;
+
+                throw new LicenseLimitException(LimitType.AdditionalAssembliesFromNuGet, "Your license doesn't support Additional Assemblies From NuGet feature.");
+
+            case nameof(UpdatePeriodicBackupCommand):
+                if (CanAssertLicenseLimits(context, minBuildVersion: MinBuildVersion54118) == false)
+                    return;
+
+                var backupTypes = LicenseManager.GetBackupTypes(databaseRecord.PeriodicBackups);
+
+                if (backupTypes.HasSnapshotBackup)
```

Comment:
what about backward compatibility? you modify one specific backup, but assert all existing backups?

Reply:
or if I had enterprise and now downgraded to professional

Reply:
If you have multiple limits in the back you will throw only for 1. First for snapshot then cloud and encrypted.
For old version (before 54118) , we don't check that. 
If we change one backup type, the rest should pass because they passed before the change.
If we downgrade to professional and don't have snapshop and change cloud backup it won't throw

Reply:
1. Existing user has snapshot backup - everything is working for him
2. Snapshot backup is only for enterprise license
3. He downgraded the license to be professional 
4. So far everything is working for him, no exceptions.
5. He upgraded the server
6. He tries to create _normal_ backup which his license supports

What would happened?

Reply:
Until now we already check license limit when downgrading so step 3 will fail.
if we do :
1. Previous server - professional license - import database with snapshot
2. Upgrade the server
3. Create normal backup 
Then we will find out that he use snapshot with  professional  an error on the snapshot.
@grisha-kotler Do we want to throw in this case or only if he change the snapshot?

Reply:
ping